common/persistence/visibility/store/sql/visibility_store.go

```diff
@@ -647,67 +776,154 @@ func (s *VisibilityStore) rowToInfo(
 	chasmMapper *chasm.VisibilitySearchAttributesMapper,
-) (*commonpb.SearchAttributes, error) {
+) (map[string]interface{}, *commonpb.SearchAttributes, error) {
 	saTypeMap, err := s.searchAttributesProvider.GetSearchAttributes(
 		s.GetIndexName(),
 		false,
 	)
 	if err != nil {
-		return nil, serviceerror.NewUnavailable(
+		return nil, nil, serviceerror.NewUnavailable(
 			fmt.Sprintf("Unable to read search attributes types: %v", err))
 	}
 
-	combinedTypeMap := make(map[string]enumspb.IndexedValueType)
-	for k, v := range saTypeMap.Custom() {
-		combinedTypeMap[k] = v
-	}
+	// Get registered CHASM SAs for this archetype (if it's a CHASM query)
+	var chasmTypeMap map[string]enumspb.IndexedValueType
 	if chasmMapper != nil {
-		chasmTypeMap := chasmMapper.SATypeMap()
-		for fieldName, fieldType := range chasmTypeMap {
-			combinedTypeMap[fieldName] = fieldType
-		}
+		chasmTypeMap = chasmMapper.SATypeMap()
 	}
-	finalTypeMap := searchattribute.NewNameTypeMap(combinedTypeMap)
 
-	// In SQLite, keyword list can return a string when there's only one element.
-	// This changes it into a slice.
+	customTypeMap := saTypeMap.Custom()
+
+	// Split row search attributes into CHASM and custom based on field name pattern
+	chasmRowSAs := make(sqlplugin.VisibilitySearchAttributes)
+	customRowSAs := make(sqlplugin.VisibilitySearchAttributes)
+
 	for name, value := range rowSearchAttributes {
-		tp, err := finalTypeMap.GetType(name)
-		if err != nil {
-			return nil, err
+		if defs.IsChasmSearchAttribute(name) {
+			// For CHASM queries, only include registered CHASM SAs
+			if chasmMapper != nil {
+				if _, isRegistered := chasmTypeMap[name]; isRegistered {
+					chasmRowSAs[name] = value
+				}
+			}
+			// For workflow queries, skip CHASM SAs entirely
+		} else if _, isCustomSA := customTypeMap[name]; isCustomSA {
+			customRowSAs[name] = value
 		}
-		if tp == enumspb.INDEXED_VALUE_TYPE_KEYWORD_LIST {
-			switch v := value.(type) {
-			case []string:
-				// no-op
-			case string:
-				(rowSearchAttributes)[name] = []string{v}
-			default:
-				return nil, serviceerror.NewInternal(
-					fmt.Sprintf("Unexpected data type for keyword list: %T (expected list of strings)", v),
-				)
+		// Ignore attributes that are neither CHASM nor custom (shouldn't happen)
+	}
+
+	// Process CHASM search attributes
+	var chasmSAs map[string]interface{}
+	if len(chasmRowSAs) > 0 {
+		chasmNameTypeMap := searchattribute.NewNameTypeMap(chasmTypeMap)
+
+		// Fix SQLite keyword list handling
+		for name, value := range chasmRowSAs {
+			tp, err := chasmNameTypeMap.GetType(name)
+			if err != nil {
+				return nil, nil, err
+			}
+			if tp == enumspb.INDEXED_VALUE_TYPE_KEYWORD_LIST {
+				switch v := value.(type) {
+				case []string:
+					// no-op
+				case string:
+					chasmRowSAs[name] = []string{v}
+				default:
+					return nil, nil, serviceerror.NewInternal(
+						fmt.Sprintf("Unexpected data type for keyword list: %T (expected list of strings)", v),
+					)
+				}
```

Comment:
You have to do this for both chasm and custom search attribute. Move to the first for-loop.